src/OmniSharp.Roslyn/OmniSharpWorkspace.cs

```diff
@@ -80,9 +83,49 @@ public void RemoveMetadataReference(ProjectId projectId, MetadataReference metad
 
         public void AddDocument(DocumentInfo documentInfo)
         {
+            // if the file has already been added as a misc file,
+            // because of a possible race condition between the updates of the project systems,
+            // remove the misc file and add the document as required
+            TryRemoveMiscellaneousDocument(documentInfo.FilePath);
+
             OnDocumentAdded(documentInfo);
         }
 
+        public DocumentId TryAddMiscellaneousDocument(string filePath, string language)
+        {
+            if (GetDocument(filePath) != null)
+                return null; //if the workspace already knows about this document then it is not a miscellaneous document
+
+            var projectInfo = miscDocumentsProjectInfos.GetOrAdd(language, CreateMiscFilesProject(language));
+            var documentId = AddDocument(projectInfo.Id, filePath);
+            return documentId;
+        }
+
+        public bool TryRemoveMiscellaneousDocument(string filePath)
+        {
+            var documentId = GetDocumentId(filePath);
+            if (documentId == null || !IsMiscellaneousDocument(documentId))
+                return false;
+
+            RemoveDocument(documentId);
+            return true;
+        }
+
+        private ProjectInfo CreateMiscFilesProject(string language)
+        {
+            string assemblyName = Guid.NewGuid().ToString("N");
+            var projectInfo = ProjectInfo.Create(
+                   id: ProjectId.CreateNewId(),
+                   version: VersionStamp.Create(),
+                   name: "MiscellaneousFiles",
+                   metadataReferences: new MetadataReference[] { MetadataReference.CreateFromFile((typeof(object).Assembly).Location) },
```

Comment:
scripting project system does something similar, where it adds some default CLR references. This is the code https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Script/ScriptContextProvider.cs#L120-L128 that ensures the mandatory "common" types are loaded and also work on Mono (where you have System.Runtime etc). Perhaps you want to use that here?